### PR TITLE
Save and update sub-playlists to users account

### PR DESCRIPTION
**Pull request by thenavz6**
## Context

[Corresponding Pivotal Tracker story](https://www.pivotaltracker.com/story/show/175033813)

## Description 💬

- Playlists can now be saved to users account by clicking save button
- Playlist is created in users account if it does not exist
- Playlist is updated if it already exists in users account
- Updated PlaylistObj to have two new fields, `collaborative: boolean, public: boolean`
- Reordered state hooks in SubPlaylist component to keep track of tracks in subplaylist
- Updated scope of app to include editing of user playlists

## Checklist ✅:

- [x] I assigned the story to myself on Pivotal Tracker before I started working on this
- [x] Added a link to the user story context
- [x] Added a description of my changes
- [x] Assigned at least two reviewers

### CODE REVIEW

File: src/helpers/helpers.tsx
Comment (featherbear) [inline src/helpers/helpers.tsx]: Typo in `paylist`

File: src/helpers/helpers.tsx
Comment (featherbear) [inline src/helpers/helpers.tsx]: Typo in `paylist`

**Review by featherbear — CHANGES_REQUESTED** [review]

**Comment by featherbear** [comment]
Typos at per review ^, otherwise looks good!

**Review by featherbear — DISMISSED** [review]
lgtm

File: src/components/Subplaylist.tsx
Comment (featherbear) [inline src/components/Subplaylist.tsx:104]: _dis kinda dodg ya feel?_

I'd think that the time it takes for both of the API calls should already be an adequate delay.

Comment (thenavz6) [reply src/components/Subplaylist.tsx:104]: LOL yeah I tried without the timeouts at first, but the api calls happen pretty quick. The timeouts don't block anything else so should be fine?

Comment (featherbear) [reply src/components/Subplaylist.tsx:104]: When your internet is too fast :')

Perhaps some sort of time comparison to have it as a minimum of `X` seconds, rather than `delay + X`.
But yeah, PR lgtm

**Comment by z5160345** [comment]
Maybe the save button should be disabled until the user has actually made a change

**Comment by thenavz6** [comment]
> Maybe the save button should be disabled until the user has actually made a change

**Comment by CarlinWilliamson** [comment]
Currently the saving spinner is green and ontop of the save button.
Could we make it the same blue and hide the button text? Or maybe just hiding the button would look best?

**Review by zacharygilmour17 — COMMENTED** [review]
Following scenario breaks application:

- make subplaylist on splitify
- delete subplaylist from spotify application
- save the subplaylist again from splitify

**Review by featherbear — COMMENTED** [review]
~@thenavz6 There will be a required refactor of the code base after https://github.com/Splitify/splitify/pull/18 and https://github.com/Splitify/splitify/pull/20 
Could you hold off on this PR until those have been merged~

Unblocked

**Review by CarlinWilliamson — CHANGES_REQUESTED** [review]
Looks pretty good.

A couple cases where it doesn't do what I'd expect
- If you change the name of the subplaylist you can't save straight away
- If you restrict by an audio feature you can't save straight away

When is the green save button supposed to appear? It flashes then gets replaced by the grey or blue one very quickly

**Comment by CarlinWilliamson** [comment]
Also getting a 400 when trying to add many songs to an existing sub-playlist

Says:
> message: "you can add a maximum of 100 tracks per request"
> status: 400

This put that subplaylist into an "unsaveable" state. The spinner just kept going forever

**Comment by CarlinWilliamson** [comment]
I just tried to save 1400 tracks. There were a couple 500 errors and not all of the tracks were saved

**Comment by featherbear** [comment]
> 
> 
> I just tried to save 1400 tracks. There were a couple 500 errors and not all of the tracks were saved

weird flex but okay

**Review by z5160345 — DISMISSED** [review]
lgtm. Good work, finally feels like a real product now that its actually adding new playlists to my spotify account

**Review by CarlinWilliamson — APPROVED** [review]
lgtm. Long time coming XD

**Review by featherbear — APPROVED** [review]
lgtm

**Pull request merged**
